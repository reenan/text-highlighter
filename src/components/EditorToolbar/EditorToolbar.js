import React, { Component } from 'react';

import { EditorState } from 'draft-js';
import './EditorToolbar.scss';


export default class EditorToolbar extends Component {

  static defaultProps = {
    editorState: EditorState.createEmpty(),
    setStyle: {},
    showToolbar: false,
    toolbarPosition: {},
    styleMap: {},
  }

  constructor(props) {
    super(props);

    this.state = {
      shouldRender: false
    }
  }

  componentDidUpdate = (prevProps) => {
    if (prevProps.showToolbar === true && this.props.showToolbar === false) {
      this.removeElementTimeout = setTimeout(() => {
        this.setState({
          shouldRender: false
        })
      }, 2000);
    } else if (this.props.showToolbar === true && this.state.shouldRender === false) {
      clearInterval(this.removeElementTimeout);

      this.setState({
        shouldRender: true
      });
    }
  }

  render() {
    const { editorState, setStyle, showToolbar, toolbarPosition, styleMap } = this.props;
    let currentStyle = editorState.getCurrentInlineStyle();

    return (
      this.state.shouldRender ?
        <div className={`toolbar ${showToolbar ? 'open' : ''}`} style={toolbarPosition}>
          {Object.keys(styleMap).map((type) =>
            <StyleButton
              key={type}
              active={currentStyle.has(type)}
              onClick={setStyle}
              style={type}
              backgroundColor={styleMap[type].backgroundColor}
            />
          )}
        </div> : null
    )
  }
}

export class StyleButton extends React.Component {

  static defaultProps = {
    active: false,
    style: null,
    onClick: () => {},
    backgroundColor: '',
  }

  onClick = (e) => {
    e.preventDefault();

    const { active, style } = this.props;
    this.props.onClick(active ? null : style);
  }

  render() {
    const { active, backgroundColor } = this.props;
    const className = `style-button ${active ? 'active' : ''}`;
    return (
      <span className={className} style={{backgroundColor}} onMouseDown={this.onClick} />
    );
  }
}