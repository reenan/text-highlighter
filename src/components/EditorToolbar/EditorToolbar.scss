@keyframes fade-in {
  0% {
    opacity: 0;
  }
  
  50% {
    opacity: 1;
  }
}

.toolbar {
  animation: fade-in 400ms ease-out;
  transition: all 200ms ease-out;
  position: fixed;
  box-shadow: 0px 0 10px -5px rgba(black, 0.5);
  background-color: white;
  z-index: 2;
  border-radius: 15px;
  padding: 8px;
  pointer-events: none;

  .style-button {
    position: relative;
    pointer-events: auto;
    float: left;
    padding: 12px;
    border-radius: 100%;
    cursor: pointer;
    opacity: 0.7;
    transition: all 400ms ease-out;
    border: 3px solid transparent;

    &:hover {
      opacity: 1;
    }

    & + .style-button {
      margin-left: 5px;
    }

    &.red {
      background-color: #f95555;
    }

    &.yellow {
      background-color: #ecec00;
    }

    &.green {
      background-color: #34c534;
    }

    &:before {
      content: '';
      transition: all 400ms ease-out;
      position: absolute;
      top: -3px;
      right: -3px;
      bottom: -3px;
      left: -3px;
      border: 2px solid transparent;
      border-radius: 100%;
    }

    &.active {
      border-color: white;
      &:before {
        border-color: #d0d0d0;
      }
    }
  }

  &:not(.open) {
    opacity: 0;

    .style-button {
      pointer-events: none;
    }
  }
}