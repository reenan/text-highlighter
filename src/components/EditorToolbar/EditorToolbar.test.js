import React from 'react';
import ReactDOM from 'react-dom';
import EditorToolbar, { StyleButton } from './EditorToolbar';
import renderer from 'react-test-renderer';
import { render } from 'react-testing-library';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<EditorToolbar />, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('should render properly with shouldRender true/false and showToolbar true/false', () => {
  let component = renderer.create(<EditorToolbar />);
  let tree = component.toJSON();

  expect(tree).toMatchSnapshot();

  component.getInstance().setState({
    shouldRender: true
  })

  tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  component = renderer.create(<EditorToolbar showToolbar={true} />);
  component.getInstance().setState({
    shouldRender: true
  })

  tree = component.toJSON();
  expect(tree).toMatchSnapshot();

});

it('should update shouldRender based on received props without crashing', (done) => {
  const { rerender } = render(<EditorToolbar />)
  rerender(<EditorToolbar showToolbar={true} />);
  rerender(<EditorToolbar showToolbar={false} />);

  setTimeout(() => {
    done();
  }, 2000);
})

it('should render style button without crashing with active true/false', () => {
  let component = renderer.create(<StyleButton />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  component = renderer.create(<StyleButton active />);
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
})

it('should be able to click on style button without crashing', () => {
  let component = renderer.create(<StyleButton />);
  component.getInstance().onClick(new Event(null));
})

it('should be able to click on style button with active without crashing', () => {
  let component = renderer.create(<StyleButton active />);
  component.getInstance().onClick(new Event(null));
})


