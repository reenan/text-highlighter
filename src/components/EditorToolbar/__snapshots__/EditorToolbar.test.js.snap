// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render properly with shouldRender true/false and showToolbar true/false 1`] = `null`;

exports[`should render properly with shouldRender true/false and showToolbar true/false 2`] = `
<div
  className="toolbar "
  style={Object {}}
/>
`;

exports[`should render properly with shouldRender true/false and showToolbar true/false 3`] = `
<div
  className="toolbar open"
  style={Object {}}
/>
`;

exports[`should render style button without crashing with active true/false 1`] = `
<span
  className="style-button "
  onMouseDown={[Function]}
  style={
    Object {
      "backgroundColor": "",
    }
  }
/>
`;

exports[`should render style button without crashing with active true/false 2`] = `
<span
  className="style-button active"
  onMouseDown={[Function]}
  style={
    Object {
      "backgroundColor": "",
    }
  }
/>
`;
